### PR TITLE
[HttpClient] simplify retry mechanism around RetryStrategyInterface

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Replaces #38466

I feel like the mechanism behind RetryableHttpClient is too bloated for no pragmatic reasons.

I propose to merge RetryDeciderInterface and RetryBackoffInterface into one single RetryStrategyInterface.

Having two separate interfaces supports no real-world use case. The only implementations we provide are trivial, and they can be reused by extending the provided GenericRetryStrategy implementation if one really doesn't want to duplicate that.

The methods are also simplified by accepting an AsyncContext as argument. This makes the signatures shorter and allows accessing the "info" of the request (allowing to decide based on the IP of the host, etc).

/cc @jderusse

### CODE REVIEW

File: src/Symfony/Component/HttpClient/Retry/GenericRetryStrategy.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+
+/**
+ * Decides to retry the request when HTTP status codes belong to the given list of codes.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class GenericRetryStrategy implements RetryStrategyInterface
+{
+    public const DEFAULT_RETRY_METHODS = [423, 425, 429, 500, 502, 503, 504, 507, 510];
```
Comment (jderusse) [inline src/Symfony/Component/HttpClient/Retry/GenericRetryStrategy.php]: `DEFAULT_RETRY_STATUSES`

File: src/Symfony/Component/HttpClient/Retry/RetryStrategyInterface.php [hunk 2]
```diff
@@ -23,5 +27,10 @@ interface RetryDeciderInterface
      *
      * @return ?bool Returns null to signal that the body is required to take a decision
      */
-    public function shouldRetry(string $requestMethod, string $requestUrl, array $requestOptions, int $responseStatusCode, array $responseHeaders, ?string $responseContent): ?bool;
+    public function shouldRetry(AsyncContext $context, ?string $responseContent, ?TransportExceptionInterface $exception): ?bool;
```
Comment (jderusse) [inline src/Symfony/Component/HttpClient/Retry/RetryStrategyInterface.php:30]: `TransportExceptionInterface $exception` is always null from `RetryableHttpClient` implementation

**Review by jderusse — COMMENTED** [review]
I like the new methods signatures.

Don't forget to update the `FrameworkExtension`

File: src/Symfony/Component/HttpClient/Retry/RetryStrategyInterface.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Retry/RetryStrategyInterface.php:30]: Sure, this anticipates #38426

**Review by jderusse — APPROVED** [review]
Thank you @nicolas-grekas. It's simpler like this.

File: src/Symfony/Component/HttpClient/Retry/GenericRetryStrategy.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Retry;
+
+use Symfony\Component\HttpClient\Response\AsyncContext;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+
+/**
+ * Decides to retry the request when HTTP status codes belong to the given list of codes.
```
Comment (stof) [inline src/Symfony/Component/HttpClient/Retry/GenericRetryStrategy.php:18]: This should document that its delay implementation is an exponential backoff. That got lost when naming it `generic`.

**Pull request merged**